### DISCUSSION TITLE
Saleor V3 Emails - handle translations

### DISCUSSION THREAD

**Discussion by JannikZed** [saleor/saleor#7285] [General] [answered]
I was just analyzing the new email plugin that most likely a lot of people are awaiting patiently. I really like the modularity and the sendgrid integration. My question is now (as the documentation is missing) - is there already a plan to use translations for text fragments? And where is the "redirect_url" set used to send out a link to the order?

In general it would be really great to get an overview of all the used handlebar variables :)

**Comment by korycins** [comment, marked as answer]
Here you have docs. There is an overview for all handlebars helper functions and payload which email plugin recieves.
https://docs.saleor.io/docs/3.0/developer/available-plugins/user-emails
Each payload has the field `language_code` which can be determined to correct user language. Some mutation inputs have new field language_code. This is the info from the changelog:

>   Add field languageCode to types: AccountInput, AccountRegisterInput, CheckoutCreateInput, CustomerInput, Order, User. Add field languageCodeEnum to Order type. Add new mutation CheckoutLanguageCodeUpdate.

So you can control the variable which defines the user's language.  I believe we don't have docs for Sendgrid yet but the payload looks the same as for UserEmailPlugin. 
AFAIK  we don't have a plan to use translations for text fragments as this is now something configurable from the dashboard. 

Also in a few days, we will introduce plugins per channel feature. By using it you will be able to build different email configurations for each channel. Assuming that you have online store for channels: DE and PL. Each of them will be able to have different template/sendgrid configuration

> And where is the "redirect_url" set used to send out a link to the order?


CheckoutComplete has argument `redirect_url`. This argument is stored in the checkout object and then passed to the order objects.
There is a validation before saving it so make sure that your domain is whitelisted -  `saleor.core.utils.url.validate_storefront_url`

**Reply by JannikZed** [reply]
amazing! Thank you very much for making that all clear. I will test the E-Mail Features out. 

The redirect URL field was new for me - you are also not using it on the demo store. But that is really exactly what we need. Thank you for bringing that up as well.

**Reply by JannikZed** [reply]
hey @korycins - I was trying to use the sendgrid feature, but it seems like the plugin is not sending the recipient address to Sendgrid: 
<img width="1212" alt="Screenshot 2021-05-07 at 18 44 14" src="https://user-images.githubusercontent.com/7526174/117571156-87ea5e00-b0cd-11eb-98f4-44b1ffd7631a.png">

FromAddress, Dynamic template ID etc. seems to be set correctly, but the delivery fails because of the missing ToAddress.

**Reply by korycins** [reply]
Hey @JannikZed  Which event are you trying to send?
The `send_email` function assigns the recipient_address:
```
def send_email(configuration: SendgridConfiguration, template_id, payload):
    recipient_email = payload["recipient_email"]
    sendgrid_client = SendGridAPIClient(configuration.api_key)
    from_email = (configuration.sender_address, configuration.sender_name)
    message = Mail(from_email=from_email, to_emails=recipient_email)
    message.dynamic_template_data = payload
    message.template_id = template_id
    sendgrid_client.send(message)
```

So I believe the issue can be connected with the given type of event.

**Reply by JannikZed** [reply]
could be! It was the order confirmation email

**Reply by korycins** [reply]
I will check this

**Reply by korycins** [reply]
Thanks for letting me know

**Reply by korycins** [reply]
@JannikZed  I just double check that. I was able to activate SendGrid send and receive an email. Everything looks okay. 
To define recipient email we are using a method for order:
```
    def get_customer_email(self):
        return self.user.email if self.user else self.user_email
```
So I believe that your case could happen only when your order doesn't have an assigned customer email (which should not happen 🤔 )

let me know if you have more details that could be helpful with reproducing the issue.

**Reply by JannikZed** [reply]
hmm -- I tried out many things. Different events, creating an order as guest and as real user. But all the time, the E-Mail address is missing. In the dashboard, I can see the address connected with the order, so everything looks fine .. 
Do you have an idea, how I can debug that case?

**Reply by korycins** [reply]
Hhm 🤔 
Can you try to insert a breakpoint in `send_email` function and check the `recipient_email`? (print will be also fine)
What version of SendGrid package do you have installed?

**Reply by JannikZed** [reply]
I found the root cause.. it's not your plugin, but an error in my Handlebars logic in the subject line. Sendgrid is not showing that correctly, but just suppressing the message an visualizing it as if the E-Mail address is missing :D really strange.

**Reply by korycins** [reply]
Thanks for letting me know. Happy to hear that you solved the issue 😄

**Comment by JustCreate3D** [comment]
Hi @korycins ,
I am currently trying to translate the user emails via the handlebar `compare` function creating a kind of switch case for the `language_code` field. I am using a single channel for selling in all countries using €, but have a different language. 
For this purpose, I am creating fake orders through the dashboard. 
I found by this occasion, that when creating an order through the dashboard, there is no option to set this parameter (`languageCode`), **nor to update it**, and I found that an order's `languageCode` is auto-set to english. 
The `orderUpdate` mutation misses this parameter as well as the `draftOrderUpdate`. 
I think, that it would be nice, to have the option to set the order's language code for orders created through the dashboard. I understand that in production, the orders are created via the checkout and that there, there is this possibility. 

Please, let me know if I'm missing something, or if I should create an issue with this.

**Reply by korycins** [reply]
Hey @JustCreate3D, 
I think you're right. Would be good to have a possibility to update it for draft order. Can you create an issue for this?

Thanks for raising this
